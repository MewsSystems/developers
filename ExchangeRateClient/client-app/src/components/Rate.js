// @flow strict
import * as React from "react";
import styled from "styled-components";
import { FaCaretUp, FaCaretDown, FaSnowflake } from "react-icons/fa";
import theme from "../theme";
import type { Theme } from "../theme/types";

type Type = "decreased" | "increased" | "default";

const Container = styled.div`
  flex: 3;
  display: inline-flex;
  position: relative;
`;

const IconContainer = styled.div`
  position: absolute;
`;

const Text = styled.span`
  margin-left: 1em;
  color: ${({ type, theme }: { type: Type, theme: Theme }) =>
    ({
      increased: theme.colors.success,
      decreased: theme.colors.alert,
      default: theme.colors.white,
    }[type])};
`;

Text.defaultProps = {
  theme,
};

const DownIcon = styled(FaCaretDown)`
  color: ${({ theme }: { theme: Theme }) => theme.colors.alert};
`;

DownIcon.defaultProps = {
  theme,
};

const UpIcon = styled(FaCaretUp)`
  color: ${({ theme }: { theme: Theme }) => theme.colors.success};
`;

UpIcon.defaultProps = {
  theme,
};

const Snowflake = styled(FaSnowflake)`
  color: ${({ theme }: { theme: Theme }) => theme.colors.color2};
`;

const getRateType = (current, before): Type => {
  if (before > current) {
    return "decreased";
  }
  if (current > before) {
    return "increased";
  }
  return "default";
};

const Icon = ({ type }: { type: Type }) => {
  if (type === "decreased") return <DownIcon data-testid="icon" />;
  if (type === "increased") return <UpIcon data-testid="icon" />;
  return <Snowflake size={25} data-testid="icon" />;
};

type Props = {| +current: number, +before: number |};

export default ({ current, before }: Props) => (
  <Container>
    <IconContainer>
      <Icon type={getRateType(current, before)} />
    </IconContainer>
    <Text type={getRateType(current, before)}>{current}</Text>
  </Container>
);
