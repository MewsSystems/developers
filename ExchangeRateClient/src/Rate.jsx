// @flow
import React from 'react';
import styled from 'styled-components';
import type { CurrencyPairRate } from './types';

type Props = {
  className: string,
  rate: CurrencyPairRate,
};

const Wrapper = styled.div``;

const CurrencyPair = styled.div``;

const CurrencyCode = styled.span`
  font-size: 0.75rem;
  text-transform: uppercase;
  letter-spacing: 1px;
`;

const CurrencyDivider = styled.span`
  padding: 0 0.1rem;
`;

const CurrencyRate = styled.span`
  font-size: 1.5rem;
`;

const trend = (prev, curr) => {
  if (prev < curr) {
    return '↑';
  }

  if (prev > curr) {
    return '↓'
  }

  return '●';
}

const Rate = ({ className, rate }: Props) => {
  const [prev, curr] = rate.rate;
  const { left, right } = rate.currencyPair;

  return (
    <Wrapper className={className}>
      <CurrencyPair>
        <CurrencyCode>{left.code}</CurrencyCode>
        <CurrencyDivider>/</CurrencyDivider>
        <CurrencyCode>{right.code}</CurrencyCode>
      </CurrencyPair>
      <CurrencyRate>{curr} {trend(prev, curr)}</CurrencyRate>
    </Wrapper>
  )
};

Rate.defaultProps = {
  className: '',
};

export default Rate;
