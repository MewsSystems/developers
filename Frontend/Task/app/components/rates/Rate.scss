@import '../../styles/mixins';
@import '../../styles/variables';
@import '../../styles/colors';

.Rate {
  @include Card();
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: repeat(auto-fill, minmax($cardItemSize, 1fr));

  &__displayName {
    font-weight: bold;
  }

  &__trend {
    &--growing {
      color: $green-500;
    }

    &--stagnating{
      color: $orange-500;
    }

    &--declining {
      color: $red-500;
    }
  }
}