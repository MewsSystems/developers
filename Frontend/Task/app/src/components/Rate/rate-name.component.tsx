import React from 'react'
import './styles.module.css'

const RateName = ({name, code}) => {
  return (
    <>
    <td scope="col">{name}</td>
    <td scope="col">{code}</td>
    </>
  )
}

export default RateName