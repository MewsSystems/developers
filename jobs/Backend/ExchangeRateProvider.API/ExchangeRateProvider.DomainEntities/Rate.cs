namespace ExchangeRateProvider.DomainEntities
{
    public class Rate
    {
        public List<CurrencyPair> Rates { get; set; }
    }
}
