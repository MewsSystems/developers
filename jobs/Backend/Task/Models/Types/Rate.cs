namespace ExchangeRateUpdater.Models.Types;

/// <summary>
/// Represents a rate with a decimal value.
/// </summary>
internal record Rate(decimal Value);