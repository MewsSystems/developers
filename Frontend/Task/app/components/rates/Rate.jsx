import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import './Rate.scss';

const Rate = (props) => {
  const trendClassName = classnames('Rate__trend', {
    'Rate__trend--stagnating': props.trend === 'stagnating',
    'Rate__trend--growing': props.trend === 'growing',
    'Rate__trend--declining': props.trend === 'declining',
  });

  return (
    <div className="Rate">
      <label className="Rate__displayName">{props.displayName}</label>
      <div className="Rate__currentRate">{props.currentRate || '0.0'}</div>
      <div className={trendClassName}>{props.trend || '~'}</div>
    </div>
  );
};

Rate.propTypes = {
  currentRate: PropTypes.number,
  displayName: PropTypes.string,
  trend: PropTypes.string
}

export default Rate;
