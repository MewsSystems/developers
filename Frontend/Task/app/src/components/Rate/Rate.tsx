import React from 'react';
import styled from 'styled-components'
import * as skin from './Rate.skin'
import {connect} from 'react-redux';

import PairName from "../PairName";
import Preloader from '../Preloader'

const RateWrapper = styled.div`${skin.RateWrapper}`;
const RateValues = styled.div`${skin.RateValues}`;
const RatesLoading = styled(Preloader)`${skin.RatesLoading}`;
const RatesPairName = styled(PairName)`${skin.RatePairName}`;


type RateProps = {
    id: string,
    pair: Array<{
        code: string,
        name: string
    }>
    rates: {
        value: number,
        trend: string
    }
    loading: boolean
}

const Rate: React.FC<RateProps> = (props) => {
    const { rates, pair, loading } = props;
    let Trend;

    switch(rates.trend) {
        case "growing":
            Trend = styled.span`${skin.Growing}`;
            break;
        case "falling":
            Trend = styled.span`${skin.Falling}`;
            break;
        default:
            Trend = styled.span`${skin.Arrow}`;
            break;
    }

    return (
        <RateWrapper>
            <RatesPairName>
                {pair}
            </RatesPairName>
            {loading && (
                <RatesLoading/>
            )}
            {!loading && (
                <RateValues>
                    {rates.value}
                    <Trend />
                </RateValues>
            )}
        </RateWrapper>
    );
};

const mapStateToProps = state => {
    return {
        loading: state.ratesLoading,
    };
};


export default connect(mapStateToProps)(Rate);