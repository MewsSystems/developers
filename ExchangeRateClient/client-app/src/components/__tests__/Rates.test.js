// @flow strict

import { render, cleanup } from "react-testing-library";
import React from "react";
import Rate from "../Rate";

afterEach(cleanup);

describe("Rates", () => {
  test("render", () => {
    const { baseElement } = render(<Rate current={10} before={23} />);
    expect(baseElement).toMatchSnapshot();
  });

  test("if current and before values are same should render the Snowflake icon", () => {
    const { getByTestId } = render(<Rate current={10} before={10} />);
    expect(getByTestId(`icon`)).toMatchSnapshot();
  });

  test("if the current is more than the before value", () => {
    const { getByTestId } = render(<Rate current={11} before={10} />);
    expect(getByTestId(`icon`)).toMatchSnapshot();
  });

  test("if the before is more than the current value", () => {
    const { getByTestId } = render(<Rate current={1} before={10} />);
    expect(getByTestId("icon")).toMatchSnapshot();
  });
});
