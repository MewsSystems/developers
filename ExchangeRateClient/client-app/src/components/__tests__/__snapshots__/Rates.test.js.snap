// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rates if current and before values are same should render the Snowflake icon 1`] = `
<svg
  stroke="currentColor"
  fill="currentColor"
  stroke-width="0"
  viewBox="0 0 448 512"
  attr="[object Object]"
  data-testid="icon"
  height="1em"
  width="1em"
>
  <path
    d="M444.816 301.639a24.12 24.12 0 0 0 2.661-16.978c-2.725-12.966-15.339-21.245-28.174-18.492l-87.407 25.046L264 256l67.896-35.215 87.407 25.046c12.835 2.753 25.449-5.526 28.174-18.492 2.725-12.966-5.471-25.708-18.305-28.461l-47.477-7.137 53.077-30.956c11.363-6.627 15.257-21.306 8.696-32.785-6.561-11.479-21.091-15.412-32.454-8.785l-53.077 30.956 17.621-45.104c4.057-12.606-2.768-26.146-15.247-30.245-12.478-4.099-25.883 2.797-29.94 15.402l-22.232 88.99-60.38 35.215V144l65.175-63.945c8.778-9.852 7.987-25.027-1.766-33.894-9.753-8.867-24.775-8.068-33.552 1.784l-29.857 37.967V24c0-13.255-10.637-24-23.758-24s-23.758 10.745-23.758 24v61.912l-29.857-37.967c-8.779-9.852-23.799-10.652-33.552-1.784-9.753 8.867-10.543 24.042-1.766 33.894L200.242 144v70.431l-60.38-35.215-22.232-88.99c-4.057-12.605-17.462-19.501-29.94-15.402-12.478 4.099-19.304 17.64-15.247 30.245l17.62 45.104-53.077-30.956c-11.363-6.627-25.893-2.694-32.454 8.785s-2.667 26.157 8.696 32.785l53.077 30.956-47.477 7.137C5.993 201.634-2.203 214.375.523 227.341c2.725 12.965 15.339 21.245 28.174 18.492l87.407-25.046L184 256l-67.896 35.215-87.406-25.045c-12.835-2.753-25.449 5.526-28.174 18.492-2.725 12.967 5.47 25.708 18.305 28.461l47.477 7.137-53.077 30.956C1.866 357.843-2.027 372.521 4.533 384s21.091 15.412 32.454 8.785l53.077-30.956-17.62 45.104a24.157 24.157 0 0 0 2.022 19.428c2.831 4.953 7.416 8.909 13.224 10.816 12.478 4.099 25.883-2.797 29.94-15.402l22.232-88.99 60.38-35.215V368l-65.175 63.945c-8.778 9.852-7.987 25.027 1.766 33.894 9.754 8.868 24.774 8.068 33.552-1.784l29.857-37.967V488c0 13.255 10.637 24 23.758 24s23.758-10.745 23.758-24v-61.912l29.857 37.967A23.59 23.59 0 0 0 295.282 472a23.534 23.534 0 0 0 15.885-6.161c9.753-8.867 10.544-24.042 1.766-33.894L247.758 368v-70.431l60.38 35.215 22.232 88.99c4.057 12.605 17.462 19.501 29.94 15.402 12.479-4.099 19.304-17.64 15.247-30.245l-17.621-45.104 53.077 30.956c11.363 6.627 25.893 2.694 32.454-8.785s2.667-26.157-8.696-32.785l-53.077-30.956 47.477-7.137c6.86-1.469 12.394-5.793 15.645-11.481z"
  />
</svg>
`;

exports[`Rates if current and before values are same should render without icon 1`] = `
<body>
  <div>
    <div
      class="sc-bdVaJa gIszhG"
    >
      <div
        class="sc-bwzfXH hNcorc"
      />
      <span
        type="default"
        class="sc-htpNat jlijjq"
      >
        10
      </span>
    </div>
  </div>
</body>
`;

exports[`Rates if the before is more than the current value 1`] = `
<svg
  stroke="currentColor"
  fill="currentColor"
  stroke-width="0"
  viewBox="0 0 320 512"
  attr="[object Object]"
  data-testid="icon"
  theme="[object Object]"
  class="undefined sc-bxivhb hhpAuV"
  height="1em"
  width="1em"
>
  <path
    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
  />
</svg>
`;

exports[`Rates if the current is more than the before value 1`] = `
<svg
  stroke="currentColor"
  fill="currentColor"
  stroke-width="0"
  viewBox="0 0 320 512"
  attr="[object Object]"
  data-testid="icon"
  theme="[object Object]"
  class="undefined sc-ifAKCX fEDaeG"
  height="1em"
  width="1em"
>
  <path
    d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
  />
</svg>
`;

exports[`Rates render 1`] = `
<body>
  <div>
    <div
      class="sc-bdVaJa gIszhG"
    >
      <div
        class="sc-bwzfXH hNcorc"
      >
        <svg
          stroke="currentColor"
          fill="currentColor"
          stroke-width="0"
          viewBox="0 0 320 512"
          attr="[object Object]"
          data-testid="icon"
          theme="[object Object]"
          class="undefined sc-bxivhb hhpAuV"
          height="1em"
          width="1em"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          />
        </svg>
      </div>
      <span
        type="decreased"
        class="sc-htpNat irVKrs"
      >
        10
      </span>
    </div>
  </div>
</body>
`;
